syntax = "proto3";

package registry;

// Registry Service
// Provides access to device, patient, and threshold profile information
service Registry {
  // GetDevice: Retrieve device information by device ID
  rpc GetDevice (GetDeviceRequest) returns (GetDeviceResponse);
  
  // GetPatient: Retrieve patient information by patient ID
  rpc GetPatient (GetPatientRequest) returns (GetPatientResponse);
  
  // GetThresholdProfile: Retrieve threshold profile for a patient or device (optional)
  rpc GetThresholdProfile (GetThresholdProfileRequest) returns (GetThresholdProfileResponse);
}

// Request to get device information
message GetDeviceRequest {
  // Version of the request schema
  string version = 1;
  
  // Device identifier
  string device_id = 2;
}

// Response with device information
message GetDeviceResponse {
  // Version of the response schema
  string version = 1;
  
  // Status of the operation
  Status status = 2;
  
  // Device information (populated if status is SUCCESS)
  Device device = 3;
  
  // ISO 8601 timestamp of the response
  string timestamp = 4;
}

// Device information
message Device {
  // Device identifier
  string device_id = 1;
  
  // Device type (e.g., "blood_pressure_monitor", "pulse_oximeter")
  string device_type = 2;
  
  // Patient ID associated with this device
  string patient_id = 3;
  
  // Device status
  DeviceStatus status = 4;
  
  // Device metadata
  map<string, string> metadata = 5;
  
  // ISO 8601 timestamp when device was registered
  string registered_at = 6;
  
  // ISO 8601 timestamp of last update
  string updated_at = 7;
}

// Device status enumeration
enum DeviceStatus {
  DEVICE_STATUS_UNSPECIFIED = 0;
  DEVICE_STATUS_ACTIVE = 1;
  DEVICE_STATUS_INACTIVE = 2;
  DEVICE_STATUS_MAINTENANCE = 3;
  DEVICE_STATUS_DECOMMISSIONED = 4;
}

// Request to get patient information
message GetPatientRequest {
  // Version of the request schema
  string version = 1;
  
  // Patient identifier
  string patient_id = 2;
}

// Response with patient information
message GetPatientResponse {
  // Version of the response schema
  string version = 1;
  
  // Status of the operation
  Status status = 2;
  
  // Patient information (populated if status is SUCCESS)
  Patient patient = 3;
  
  // ISO 8601 timestamp of the response
  string timestamp = 4;
}

// Patient information
message Patient {
  // Patient identifier
  string patient_id = 1;
  
  // Patient age in years
  int32 age = 2;
  
  // Patient gender
  Gender gender = 3;
  
  // List of medical conditions
  repeated string medical_conditions = 4;
  
  // List of current medications
  repeated string medications = 5;
  
  // List of known allergies
  repeated string allergies = 6;
  
  // Patient metadata
  map<string, string> metadata = 7;
  
  // ISO 8601 timestamp when patient was registered
  string registered_at = 8;
  
  // ISO 8601 timestamp of last update
  string updated_at = 9;
}

// Gender enumeration
enum Gender {
  GENDER_UNSPECIFIED = 0;
  GENDER_MALE = 1;
  GENDER_FEMALE = 2;
  GENDER_OTHER = 3;
}

// Request to get threshold profile
message GetThresholdProfileRequest {
  // Version of the request schema
  string version = 1;
  
  // Patient identifier (required)
  string patient_id = 2;
  
  // Optional device identifier (if thresholds are device-specific)
  string device_id = 3;
}

// Response with threshold profile
message GetThresholdProfileResponse {
  // Version of the response schema
  string version = 1;
  
  // Status of the operation
  Status status = 2;
  
  // Threshold profile (populated if status is SUCCESS)
  ThresholdProfile profile = 3;
  
  // ISO 8601 timestamp of the response
  string timestamp = 4;
}

// Threshold profile for vital signs
message ThresholdProfile {
  // Profile identifier
  string profile_id = 1;
  
  // Patient ID this profile applies to
  string patient_id = 2;
  
  // Optional device ID if profile is device-specific
  string device_id = 3;
  
  // Heart rate thresholds
  VitalThreshold heart_rate = 4;
  
  // Blood pressure thresholds
  BloodPressureThreshold blood_pressure = 5;
  
  // Temperature thresholds
  VitalThreshold temperature = 6;
  
  // Oxygen saturation thresholds
  VitalThreshold oxygen_saturation = 7;
  
  // Respiratory rate thresholds
  VitalThreshold respiratory_rate = 8;
  
  // ISO 8601 timestamp when profile was created
  string created_at = 9;
  
  // ISO 8601 timestamp of last update
  string updated_at = 10;
}

// Threshold for a single vital sign
message VitalThreshold {
  // Lower bound threshold
  double lower_bound = 1;
  
  // Upper bound threshold
  double upper_bound = 2;
  
  // Unit of measurement
  string unit = 3;
  
  // Critical lower bound (triggers critical alert)
  double critical_lower = 4;
  
  // Critical upper bound (triggers critical alert)
  double critical_upper = 5;
}

// Blood pressure thresholds (systolic and diastolic)
message BloodPressureThreshold {
  // Systolic pressure thresholds
  VitalThreshold systolic = 1;
  
  // Diastolic pressure thresholds
  VitalThreshold diastolic = 2;
  
  // Unit (typically "mmHg")
  string unit = 3;
}

// Status enumeration
enum Status {
  STATUS_UNSPECIFIED = 0;
  STATUS_SUCCESS = 1;
  STATUS_NOT_FOUND = 2;
  STATUS_INVALID_REQUEST = 3;
  STATUS_INTERNAL_ERROR = 4;
}

